### PR TITLE
Simplify checksum caching code.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Metadata.cs [hunk 1]
```diff
@@ -208,30 +208,17 @@ private static string GetMetadataNameWithoutBackticks(MetadataReader reader, Str
         public static Checksum GetMetadataChecksum(
             SolutionServices services, PortableExecutableReference reference, CancellationToken cancellationToken)
         {
-            // We can reuse the index for any given reference as long as it hasn't changed.
-            // So our checksum is just the checksum for the PEReference itself.
-            // First see if the value is already in the cache, to avoid an allocation if possible.
-            if (ChecksumCache.TryGetValue(reference, out var cached))
+            return ChecksumCache.GetOrCreate(reference, static (reference, tuple) =>
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Metadata.cs:211]: inlined this.  a static lambda ensures we're not allocating.

File: src/Workspaces/Core/Portable/Workspace/Solution/DocumentInfo.cs [hunk 2]
```diff
@@ -154,7 +154,7 @@ internal sealed class DocumentAttributes(
             SourceCodeKind sourceCodeKind,
             string? filePath,
             bool isGenerated,
-            bool designTimeOnly) : IChecksummedObject, IObjectWritable
+            bool designTimeOnly) : IObjectWritable
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/DocumentInfo.cs:157]: thse types implmeented this interface, just to pass to a helper which would check for the interface, then call back into .Checksum.

Remove this and just updated those places to access the checksum directly.

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectState_Checksum.cs [hunk 3]
```diff
@@ -53,16 +55,18 @@ private async Task<ProjectStateChecksums> ComputeChecksumsAsync(CancellationToke
 
                     var serializer = LanguageServices.SolutionServices.GetService<ISerializerService>();
 
-                    var infoChecksum = serializer.CreateChecksum(ProjectInfo.Attributes, cancellationToken);
+                    var infoChecksum = this.ProjectInfo.Attributes.Checksum;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/ProjectState_Checksum.cs:58]: no need for double indirection.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState_Checksum.cs [hunk 4]
```diff
@@ -21,6 +21,11 @@ namespace Microsoft.CodeAnalysis
 {
     internal partial class SolutionState
     {
+        private static readonly ConditionalWeakTable<IReadOnlyList<ProjectId>, IReadOnlyList<ProjectId>> s_projectIdToSortedProjectsMap = new();
+
+        public static IReadOnlyList<ProjectId> GetOrCreateSortedProjectIds(IReadOnlyList<ProjectId> unorderedList)
+            => s_projectIdToSortedProjectsMap.GetValue(unorderedList, projectIds => projectIds.OrderBy(id => id.Id).ToImmutableArray());
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState_Checksum.cs:27]: this is jsut a move.  from a common location, to the place where it is needed.

File: src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs [hunk 5]
```diff
@@ -376,33 +378,43 @@ public static DocumentStateChecksums Deserialize(ObjectReader reader)
 /// </summary>
 internal static class ChecksumCache
 {
-    private static readonly ConditionalWeakTable<object, object> s_cache = new();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs:379]: updated this type to be strongly typed.  no object->object maps.

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectState_Checksum.cs [hunk 6]
```diff
@@ -53,16 +55,18 @@ private async Task<ProjectStateChecksums> ComputeChecksumsAsync(CancellationToke
...
 
-                    var projectReferenceChecksums = ChecksumCache.GetOrCreate<ChecksumCollection>(ProjectReferences, _ => new ChecksumCollection(ProjectReferences.SelectAsArray(r => serializer.CreateChecksum(r, cancellationToken))));
-                    var metadataReferenceChecksums = ChecksumCache.GetOrCreate<ChecksumCollection>(MetadataReferences, _ => new ChecksumCollection(MetadataReferences.SelectAsArray(r => serializer.CreateChecksum(r, cancellationToken))));
-                    var analyzerReferenceChecksums = ChecksumCache.GetOrCreate<ChecksumCollection>(AnalyzerReferences, _ => new ChecksumCollection(AnalyzerReferences.SelectAsArray(r => serializer.CreateChecksum(r, cancellationToken))));
+                    var projectReferenceChecksums = ChecksumCache.GetOrCreateChecksumCollection(ProjectReferences, serializer, cancellationToken);
+                    var metadataReferenceChecksums = ChecksumCache.GetOrCreateChecksumCollection(MetadataReferences, serializer, cancellationToken);
+                    var analyzerReferenceChecksums = ChecksumCache.GetOrCreateChecksumCollection(AnalyzerReferences, serializer, cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/ProjectState_Checksum.cs:69]: common patterns moved into ChecksumCache.

File: src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs:379]: the diff here is terrible.  i recommend just looking at the 'after'.

**Comment by CyrusNajmabadi** [comment]
@ToddGrun this is ready for review.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState_Checksum.cs (hunk 4 above)
Comment (ToddGrun) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState_Checksum.cs:27]: Where is this being moved from?

Comment (ToddGrun) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState_Checksum.cs:27]: nm. found it!

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Pull request merged**